store/tikv/lock_resolver.go

```diff
@@ -578,10 +595,258 @@ func (lr *LockResolver) getTxnStatus(bo *Backoffer, txnID uint64, primary []byte
+			}
+			data.commitTs = commitTS
+		}
+		data.missingLock = true
+
+		if data.commitTs != commitTS {
+			return errors.Errorf("commit TS mismatch in async commit recovery: %v and %v", data.commitTs, commitTS)
+		}
+
+		// We do not need to resolve the remaining locks because TiKV will have resolved them as appropriate.
+		return nil
+	}
+
+	logutil.BgLogger().Info("addLocks all locks present", zap.Uint64("start ts", startTS))
+	// Save all locks to be resolved.
+	var infos []*Lock
+	for _, lockInfo := range locks {
+		if lockInfo.LockVersion != startTS {
+			err := errors.Errorf("unexpected timestamp, expected: %v, found: %v", startTS, lockInfo.LockVersion)
+			logutil.BgLogger().Error("addLocks error", zap.Error(err))
+			return err
+		}
+
+		if !data.missingLock && lockInfo.MinCommitTs > data.commitTs {
+			data.commitTs = lockInfo.MinCommitTs
+		}
+		infos = append(infos, NewLock(lockInfo))
+	}
+	data.locks[curRegionID] = infos
+	return nil
+}
+
+func (lr *LockResolver) checkSecondaries(bo *Backoffer, l *Lock, curKeys [][]byte, curRegionID RegionVerID, shared *asyncResolveData) error {
+	checkReq := &kvrpcpb.CheckSecondaryLocksRequest{
+		Keys:         curKeys,
+		StartVersion: l.TxnID,
+	}
+	req := tikvrpc.NewRequest(tikvrpc.CmdCheckSecondaryLocks, checkReq)
+	tikvLockResolverCountWithQueryCheckSecondaryLocks.Inc()
+	resp, err := lr.store.SendReq(bo, req, curRegionID, readTimeoutShort)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	regionErr, err := resp.GetRegionError()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if regionErr != nil {
+		err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		logutil.BgLogger().Info("checkSecondaries region error, regrouping", zap.String("primary lock", l.String()), zap.Uint64("region", curRegionID.GetID()))
+
+		// If regions have changed, then we might need to regroup the keys. Since this should be rare and for the sake
+		// of simplicity, we will resolve regions sequentially.
+		regions, _, err := lr.store.GetRegionCache().GroupKeysByRegion(bo, curKeys, nil)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for regionID, keys := range regions {
+			// Recursion will terminate because the resolve request succeeds or the Backoffer reaches its limit.
+			if err = lr.checkSecondaries(bo, l, keys, regionID, shared); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	if resp.Resp == nil {
+		return errors.Trace(ErrBodyMissing)
+	}
+
+	checkResp := resp.Resp.(*kvrpcpb.CheckSecondaryLocksResponse)
+	return shared.addLocks(checkResp.Locks, curRegionID, len(curKeys), l.TxnID, checkResp.CommitTs)
+}
+
+// resolveLockAsync resolves l assuming it was locked using the async commit protocol.
+func (lr *LockResolver) resolveLockAsync(bo *Backoffer, l *Lock, status TxnStatus, cleanRegions map[RegionVerID]struct{}) error {
```

Comment:
`cleanRegions` isn't needed because `resolveLockAsync` resolves all locks belonging to the transaction.

Reply:
I think it's not resolved... `cleanRegions` is used to avoid sending `resolveLock` requests to the same region more than once. And `resolveLockAsync` doesn't need it because it resolves all locks of a transaction.

Reply:
SOrry, I had resolved it locally and hadn't pushed yet. Change is in the recent update